File: c++/longest-substring-without-repeating-characters.cc
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int l = 0;
        int maxSubstr = 0;
        set<char> usedChars;

        for (char r : s) {
            if (!usedChars.contains(r)) {
                usedChars.insert(r);
                maxSubstr = max(maxSubstr, (int) usedChars.size());
                continue;
            }
            while (usedChars.contains(r)) {
                usedChars.erase(s[l]);
                l++;
            }
            usedChars.insert(r);
        }
        return maxSubstr;
    }
};

File: python/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if len(s) < 2:
            return len(s)
        longestSubstring = 1
        left, right = 0, 1
        curChars = set()
        curChars.add(s[left])
        while left <= right and right < len(s):
            if s[right] in curChars:
                while (s[right] in curChars) and left <= right:
                    curChars.remove(s[left])
                    left += 1
            else:
                longestSubstring = max(longestSubstring, right - left + 1)
                curChars.add(s[right])
                right += 1

        return longestSubstring
